src/core/dense_set.h

```diff
@@ -0,0 +1,473 @@
+// Copyright 2022, Roman Gershman.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+#pragma once
+
+#include <functional>
+#include <iterator>
+#include <memory_resource>
+#include <type_traits>
+
+#include "glog/logging.h"
+
+namespace dfly {
+
+// DenseSet is a nice but over-optimized data-structure. Probably is not worth it in the first
+// place but sometimes the OCD kicks in and one can not resist.
+// The advantage of it over redis-dict is smaller meta-data waste.
+// dictEntry is 24 bytes, i.e it uses at least 32N bytes where N is the expected length.
+// dict requires to allocate dictEntry per each addition in addition to the supplied key.
+// It also wastes space in case of a set because it stores a value pointer inside dictEntry.
+// To summarize:
+// 100% utilized dict uses N*24 + N*8 = 32N bytes not including the key space.
+// for 75% utilization (1/0.75 buckets): N*1.33*8 + N*24 = 35N
+//
+// This class uses 8 bytes per bucket (similarly to dictEntry*) but it used it for both
+// links and keys. For most cases, we remove the need for another redirection layer
+// and just store the key, so no "dictEntry" allocations occur.
+// For those cells that require chaining, the bucket is
+// changed in run-time to represent a linked chain.
+// Additional feature - in order to to reduce collisions, we insert items into
+// neighbour cells but only if they are empty (not chains). This way we reduce the number of
+// empty (unused) spaces at full utilization from 36% to ~21%.
+// 100% utilized table requires: N*8 + 0.2N*16 = 11.2N bytes or ~20 bytes savings.
+// 75% utilization: N*1.33*8 + 0.12N*16 = 13N or ~22 bytes savings per record.
+// with potential replacements of hset/zset data structures.
+// static_assert(sizeof(dictEntry) == 24);
+
+class DenseSet {
+ public:
+  bool Reserve(size_t sz);
+
+ protected:
+  // Virtual functions to be implemented for generic data
+  virtual uint64_t Hash(const void*) const = 0;
```

Comment:
when you use virtual inheritance in c++ you should **always** declare/implement a virtual destructor.
the reason is: otherwise, `delete base_ptr;` will call the base class implementation without calling the derived one first.

(and I am sorry that a c++ is so broken that it allows you to compile this code).

Reply:
Is `virtual ~DenseSet();` on line 469 not a virtual destructor? Should it be moved to the top of the class declaration to be more clear?